itest/lnd_sendonion_test.go

```diff
@@ -0,0 +1,145 @@
+package itest
+
+import (
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/lightningnetwork/lnd/lnrpc"
+	"github.com/lightningnetwork/lnd/lnrpc/routerrpc"
+	"github.com/lightningnetwork/lnd/lntest"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/stretchr/testify/require"
+)
+
+func testSendOnion(ht *lntest.HarnessTest) {
```

Comment:
I'm still learning the goals here - so in this test setup, we always want to have Bob being the first hop, and the rest can be anyone?

If the goal is to ensure a payment goes through `Alice -> Bob`, I think we can do something like this instead,
1. Bob queries a route to Dave via `QueryRoutes`.
2. Alice inserts a hop in the above route, and build the full path via `BuildRoute`
3. Alice sends the payment via `SendToRouteV2`.

Reply:
I would say the goal is more to verify that Alice's node is able to correctly accept and initiate a payment using the _**SendOnion**_ RPC. This is how I have conceptualized the differences between various payment sending RPCS:

- **_SendPaymentV2_**:
  - Accepts either an invoice or a destination public key in the case of a key send.
  - The payment route is computed by the daemon.
  - Onion is computed by the daemon.
  - Caller does not have privacy from daemon operator.
- **_SendToRouteV2_**:
  - Accepts a route - an list of hops in JSON - to use for the payment.
  - Payment route is computed by the caller (implies access to channel graph).
  - Onion is computed by the daemon.
  - Caller does not have privacy from daemon operator.
- **_SendOnion_**:
  - Payment route is computed by the caller.
  - Onion is computed by the caller.
  - Caller does have privacy from daemon operator (?) The daemon just forwards the onion out the first hop!

There are a couple envisioned uses for a **_SendOnion_** style RPC that I know of:
1. If we allow the **_ChannelRouter_** to run in a separate process from the main _**lnd**_ binary, then path finding would be done by this external process (eg: lightning proxy) and then the onion delivered to _**lnd**_.
    - NOTE: This is the result of trying to re-use the _**ChannelRouter**_ code with minimal changes. It is possible that there are other ways to do this, however they may require either heavier lift/change set to the _**ChannelRouter**_ or that we avoid re-using the existing _**ChannelRouter**_ code all together.
2. A hosted node provider could run the node on behalf of it's customers. These customers could perform path finding and onion construction on their own device and then submit the onions to be sent by their node running in the hosted infrastructure. I believe this is what Core Lightning calls "oblivious sends" and they market as a way to maintain privacy from the provider in something like Green Light.